keps/sig-multicluster/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,974 @@
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+ -  Feature gate (also fill in values in `kep.yaml`)
+    *   Feature gate name: StatefulSetSlice
+    *   Components depending on the feature gate:
+        *   `kube-controller-manager`: Controls which replica ordinals are created
+        *   `kube-apiserver`: Manages the new policy field `ordinals.start`
+ -  Other
+    *   Describe the mechanism:
+    *   Will enabling / disabling the feature require downtime of the control plane? **No**
+    *   Will enabling / disabling the feature require downtime or reprovisioning of a node? **No**
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+No, if the new StatefulSet field `.spec.ordinals.start` is unset,
+StatefulSet will retain existing behavior.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes, disabling the feature gate will cause the new field to be ignored
+
+Note that if disabled, StatefulSet will implicitly number ordinals starting from `0`. This can cause churn of pods when disabled, if `ordinals.start` is not `0`. Care should be taken when disabling this feature on a spec that has `ordinals.start` set, and should only be done when pod churn or disruption can be tolerated.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+StatefulSet will see the `ordinals.start` field and scale pods to start from this ordinal. This can cause churn of pods when enabled, if `ordinals.start` is not `0`. Care should be taken when enabling this feature on a spec that has `ordinals.start` set, and should only be done when pod churn or disruption can be tolerated.
```

Comment:
With the default policy we run the risk of the controller getting stuck.  An explicit e2e test for default ordered pod management policy going from `(1,7)` to `(0,7)` should cover that sufficiently I think.  Might also want to have an unhealthy index 7 scenario in unit tests (`(1,7)` with ordinal 7 unhealthy, update start and result is ordinal 7 is deleted and 0 is created, and 7 doesn't cause ordered management policy to get stuck).

Reply:
> with ordinal 7 unhealthy, update start and result is ordinal 7 is deleted and 0 is created, and 7 doesn't cause ordered management policy to get stuck).

Isn't that in opposition to what we state in the [docs](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#forced-rollback) that broken rollout will require a manual intervention? I'd expect a similar behavior in this case as well, no?